django/thunderstore/frontend/api/experimental/tests/test_package_details.py

```diff
@@ -0,0 +1,192 @@
+from typing import Dict
+
+import pytest
+from django.urls import reverse
+from rest_framework.test import APIClient
+
+from thunderstore.community.factories import PackageListingFactory
+from thunderstore.community.models import PackageListing
+from thunderstore.repository.factories import (
+    PackageRatingFactory,
+    PackageVersionFactory,
+)
+
+
+@pytest.mark.django_db
+def test_inactive_package_returns_404(api_client: APIClient) -> None:
+    listing = PackageListingFactory(package_kwargs={"is_active": False})
+
+    response = api_client.get(__url_for_listing(listing))
+
+    assert response.status_code == 404
+
+
+@pytest.mark.django_db
+def test_inactive_package_version_returns_404(api_client: APIClient) -> None:
+    listing = PackageListingFactory(package_version_kwargs={"is_active": False})
+
+    response = api_client.get(__url_for_listing(listing))
+
+    assert response.status_code == 404
+
+
+@pytest.mark.django_db
+def test_unlisted_community_returns_404(api_client: APIClient) -> None:
+    listing = PackageListingFactory(community_kwargs={"is_listed": False})
+
+    response = api_client.get(__url_for_listing(listing))
+
+    assert response.status_code == 404
+
+
+@pytest.mark.django_db
+def test_dependant_count_is_annotated_correctly(api_client: APIClient) -> None:
+    dependency = PackageListingFactory()
+    dependency.package.latest.dependants.set(
+        [PackageVersionFactory(), PackageVersionFactory()]
+    )
+
+    data = __query_api(api_client, dependency)
+
+    assert data["dependant_count"] == 2
+
+
+@pytest.mark.django_db
+def test_downloads_count_is_annotated_correctly(api_client: APIClient) -> None:
+    listing = PackageListingFactory(package_version_kwargs={"downloads": 0})
+    PackageVersionFactory(package=listing.package, downloads=1, version_number="1.0.1")
+    PackageVersionFactory(package=listing.package, downloads=2, version_number="1.0.2")
+
+    data = __query_api(api_client, listing)
+
+    assert data["download_count"] == 3
+
+
+@pytest.mark.django_db
+def test_rating_is_annotated_correctly_when_no_ratings(api_client: APIClient) -> None:
+    listing = PackageListingFactory()
+
+    data = __query_api(api_client, listing)
+
+    assert data["rating_score"] == 0
+
+
+@pytest.mark.django_db
+def test_rating_is_annotated_correctly_when_ratings(api_client: APIClient) -> None:
```

Comment:
I'd almost suggest combining this and `test_rating_is_annotated_correctly_when_no_ratings` to a single parametrized test case, which will call `PackageRatingFactory` the amount of times defined in the parametrized decorator, and assert the rating score matches that count.

Standalone tests are fine too, but if you want to familiarize yourself with parametrized test cases, this would be a good example case.

Reply:
I'm familiar with parametrized tests, but when writing the tests I considered these two separate, "named" test cases. But I can see the argument for them not being separate too, so changing it in this instance.

Is this something that we want to have a "house rule" or project-specific way of doing things?

Reply:
I don't think it's necessary here (or should be forced), but this is a borderline case where either approach would work. Named tests can be better in some scenarios to make it easier to see what fails, as they tend to be more verbose in the names for the specific failure reason.

Primary reason I commented on it is because I hadn't seen much use of parameterized tests and I wasn't sure if that was a conscious decision or just because it wasn't a habit.

In other words, if it was a conscious decision to _not_ use a parametrized test here, but it was considered, that's completely fine, just wanted to make sure it indeed was considered.

Reply:
Well, in this case it was *not* considered 😄 In my head "stuff happens when x exists" and "other stuff happens when x doesn't exist" were so clearly different things that they should be tested in separate test cases. But as I said, I can see that "correct stuff happens with different counts of x" is a valid and logical approach as well, and as a bonus provides more than just the binary test cases. And I also implemented this already soooo...